From 3d2698669ee121a0724ea5224c0d4797e491a8d3 Mon Sep 17 00:00:00 2001
From: Sangjin Lee <sjlee@apache.org>
Date: Wed, 17 Feb 2016 20:55:21 -0800
Subject: [PATCH 1368/2748] YARN-4690. Skip object allocation in
 FSAppAttempt#getResourceUsage when possible (Ming
 Ma via sjlee)

(cherry picked from commit 7de70680fe44967e2afc92ba4c92f8e7afa7b151)
(cherry picked from commit 23b5c71729c26c2879e6cc41ce5862af0215d1fa)

Change-Id: I310656d79c2901b8b0a2b4927dadef15e8684327
---
 .../scheduler/fair/FSAppAttempt.java               |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java
index 54fc5d9..3682ea2 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java
@@ -889,7 +889,12 @@ public Resource getMaxShare() {
   public Resource getResourceUsage() {
     // Here the getPreemptedResources() always return zero, except in
     // a preemption round
-    return Resources.subtract(getCurrentConsumption(), getPreemptedResources());
+    // In the common case where preempted resource is zero, return the
+    // current consumption Resource object directly without calling
+    // Resources.subtract which creates a new Resource object for each call.
+    return getPreemptedResources().equals(Resources.none()) ?
+        getCurrentConsumption() :
+        Resources.subtract(getCurrentConsumption(), getPreemptedResources());
   }
 
   @Override
-- 
1.7.9.5

